File: 2021/dart/lib/day02.dart
import 'utils.dart' as utils;

int day02() {
  var lines = utils.readFile("../inputs/day02.txt");
  var depth = 0;
  var horizontal = 0;

  for (var line in lines) {
    var step = line.split(" ");
    var direction = step[0];
    var value = int.parse(step[1]);

    if (direction == "forward") {
      horizontal += value;
    }

    if (direction == "down") {
      depth += value;
    }

    if (direction == "up") {
      depth -= value;
    }
  }

  return depth * horizontal;
}

int day02Part2() {
  var lines = utils.readFile("../inputs/day02.txt");
  var depth = 0;
  var horizontal = 0;
  var aim = 0;

  for (var line in lines) {
    var step = line.split(" ");
    var direction = step[0];
    var value = int.parse(step[1]);

    if (direction == "forward") {
      horizontal += value;
      depth += aim * value;
    }

    if (direction == "down") {
      aim += value;
    }

    if (direction == "up") {
      aim -= value;
    }
  }

  return depth * horizontal;
}

File: 2021/rust/src/day02.rs
use crate::utils;

pub fn day02() {
    let steps: Vec<String> = utils::read_input_file("src/day02_input.txt").unwrap();
    let mut depth = 0;
    let mut horizontal = 0;

    for step in steps {
        let mut iter = step.split_whitespace();
        let direction = iter.next().unwrap();
        let value: i32 = iter.next().unwrap().parse().unwrap();

        match direction {
            "up" => depth = depth - value,
            "forward" => horizontal = horizontal + value,
            "down" => depth = depth + value,
            &_ => (),
        }
    }

    println!("{}", depth * horizontal);
}

pub fn day02_part2() {
    let steps: Vec<String> = utils::read_input_file("src/day02_input.txt").unwrap();
    let mut depth = 0;
    let mut horizontal = 0;
    let mut aim = 0;

    for step in steps {
        let mut iter = step.split_whitespace();
        let direction = iter.next().unwrap();
        let value: i32 = iter.next().unwrap().parse().unwrap();

        match direction {
            "up" => aim = aim - value,
            "down" => aim = aim + value,
            "forward" => {
                horizontal = horizontal + value;
                depth = depth + aim * value;
            }
            &_ => (),
        }
    }

    println!("{}", depth * horizontal);
}
